internal/tcp/tcp.go

```diff
@@ -228,349 +189,219 @@ var (
+		return
 	}
-	binary.Read(r, binary.BigEndian, &pkt.seqNum)
-	binary.Read(r, binary.BigEndian, &pkt.ackNum)
-
-	var mix uint16
-	binary.Read(r, binary.BigEndian, &mix)
-	pkt.dataOffset = byte(mix >> 12)  // top 4 bits
-	pkt.reserved = byte(mix >> 9 & 7) // 3 bits
-	pkt.ECN = byte(mix >> 6 & 7)      // 3 bits
-	pkt.ctrl = byte(mix & 0x3f)       // bottom 6 bits
-
-	binary.Read(r, binary.BigEndian, &pkt.window)
-	binary.Read(r, binary.BigEndian, &pkt.checksum)
-	binary.Read(r, binary.BigEndian, &pkt.urgent)
-
-	if pkt.dataOffset > 5 {
-		binary.Read(r, binary.BigEndian, &pkt.options)
+	tcpHeader = *tcp
+
+	if len(tcp.Payload) >= defines.TimestampPayloadLengthBytes {
+		ts := append([]byte(nil), tcp.Payload[:defines.TimestampPayloadLengthBytes]...)
+		err = payload.UnmarshalBinary(ts)
 	}
 
-	headerLen := pkt.dataOffset * 4
-	if len(data) > int(headerLen) {
-		ts := make([]byte, defines.TimestampPayloadLengthBytes)
-		binary.Read(r, binary.BigEndian, &ts)
+	return
+}
 
-		var unByteTime time.Time
-		if err := unByteTime.UnmarshalBinary(ts); err == nil {
-			pkt.Ts = unByteTime
+// parsePktIP parses the IP header of an incoming packet and extracts the src IP addr and DSCP value.
+func parsePktIP(pkt gopacket.Packet) (srcIP net.IP, dscpv DSCPValue, err error) {
+	switch pkt.NetworkLayer().LayerType() {
+	case layers.LayerTypeIPv4:
+		ip := pkt.Layer(layers.LayerTypeIPv4).(*layers.IPv4)
+		if ip == nil {
+			err = errors.New("Invalid IPv4 Layer")
+			return
 		}
+		srcIP = ip.SrcIP
+		dscpv = DSCPValue(ip.TOS)
+	case layers.LayerTypeIPv6:
+		ip := pkt.Layer(layers.LayerTypeIPv6).(*layers.IPv6)
+		if ip == nil {
+			err = errors.New("Invalid IPv6 Layer")
+			return
+		}
+		srcIP = ip.SrcIP
+		dscpv = DSCPValue(ip.TrafficClass)
 	}
-
-	return &pkt, true
+	return
 }
 
 // makePkt creates and serializes a TCP Echo.
 func makePkt(
-	af string,
-	srcAddr *net.IP,
-	dstAddr *net.IP,
+	af int,
+	srcAddr net.IP,
+	dstAddr net.IP,
 	srcPort uint16,
 	dstPort uint16,
-	flags uint8,
+	dscpv DSCPValue,
+	flags tcpFlags,
 	seqNum uint32,
 	ackNum uint32,
 ) ([]byte, error) {
-	var err error
-	packet := tcpPacket{
-		tcpHeader{
-			srcPort:    uint16(srcPort), // ephemeral port
-			dstPort:    uint16(dstPort),
-			seqNum:     seqNum,
-			ackNum:     ackNum,
-			dataOffset: uint8(5), // 4 bits
-			reserved:   0,        // 3 bits
-			ECN:        0,        // 3 bits
-			ctrl:       flags,    // 6 bits (000010, SYN bit set)
-			window:     0xaaaa,   // The amount of data that it is able to accept in bytes
-			checksum:   0,        // Kernel will set this if it's 0
-			urgent:     0,
-			options:    []tcpOption{},
-		},
-		tcpPayload{},
-	}
+	buf := gopacket.NewSerializeBuffer()
 
-	// When replying with SYN+ACK, a time-stamped payload is included
-	if flags&syn != 0 && flags&ack != 0 {
-		packet.tcpPayload.Ts = timeNow()
+	// gopacket serialization options for IP header are OS specific
+	optsIP := ip.GetIPLayerOptions()
+	optsTCP := gopacket.SerializeOptions{
```

Comment:
is `makePkt` called repeatedly? if so, the `optsTCP` can be a file-scope var since it doesn't change and there's no reason to allocate one each time.